### PR TITLE
Point discharge with diffusion example

### PR DESCRIPTION
**Pull request by jwallwork23**
Following the closing of issue #188, this PR acts to check that the 2D steady state tracer solver works.

Convergence of the error due to evaluating two linear functionals is assessed under uniform mesh refinement. Convergence rates for the case `hydrodynamics=False` below.

convergence_aligned
(Aligned source and receiver)

convergence_offset
(Offset source and receiver)

This could be a test, if that would be useful / there aren't too many similar tests.

### CODE REVIEW

File: examples/point_discharge/point_discharge.py [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+        self.source_r = 0.05606298 if self.use_lax_friedrichs_tracer else 0.05606298
+        self.source_value = 100.0
+
+        # Boundary conditions
+        self.boundary_conditions = {
+            'tracer': {
+                1: {'value': Constant(0.0)},      # inflow
+                2: {'open': None},                # outflow
+                3: {'diff_flux': Constant(0.0)},  # Neumann
+                4: {'diff_flux': Constant(0.0)},  # Neumann
```
Comment (tkarna) [inline examples/point_discharge/point_discharge.py:98]: I think this is equivalent to omitting the boundary condition declaration (which implies that the boundary flux term is omitted).

File: examples/point_discharge/point_discharge.py [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+
+    def quantity_of_interest(self, sol):
+        kernel = self.quantity_of_interest_kernel(sol.function_space().mesh())
+        return assemble(inner(kernel, sol)*dx(degree=12))
+
+    def analytical_quantity_of_interest(self, mesh):
+        """
+        The analytical solution can be found in [1]. Due to the modified
+        Bessel function, it cannot be evaluated exactly and instead must
+        be computed using quadrature.
```
Comment (tkarna) [inline examples/point_discharge/point_discharge.py:146]: a quadrature

File: examples/point_discharge/point_discharge.py [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+    expr2 = exp(-x)/sqrt(x)*(1.25331414 + y2*(-0.7832358e-1 + y2*(0.2189568e-1 + y2*(
+        -0.1062446e-1 + y2*(0.587872e-2 + y2*(-0.251540e-2 + y2*0.53208e-3))))))
+    return conditional(ge(x, 2), expr2, expr1)
+
+
+class PointDischargeParameters(object):
+    """
+    Problem parameter class, including point source representation.
+
+    Delta functions are not in H1 and hence do not live in the FunctionSpaces we
```
Comment (tkarna) [inline examples/point_discharge/point_discharge.py:69]: Could mention that the analytical solution involves a point source (dirac delta). And the analytical solution gives the tracer value at an arbitrary "receiver" location (if I got that right). Also the offset parameter is not explained, it moves the receiver y location to be on off-axis?

File: examples/point_discharge/point_discharge.py [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+
+    def source(self, fs):
+        triple = (self.source_x, self.source_y, self.source_r)
+        return self.gaussian(fs.mesh(), triple, scaling=self.source_value)
+
+    def bathymetry(self, fs):
+        return Function(fs).assign(5.0)
+
+    def quantity_of_interest_kernel(self, mesh):
+        triple = (20.0, 7.5 if self.offset else 5.0, 0.5)
```
Comment (tkarna) [inline examples/point_discharge/point_discharge.py:132]: minor detail: took me a while to figure out what triple is, you could define a x,y,radius dict and pass **dict to ball.

File: examples/point_discharge/point_discharge.py [hunk 5]
```diff
@@ -0,0 +1,261 @@
+"""
+TELEMAC-2D `Point Discharge with Diffusion' test case
+=====================================================
+
+Solves tracer advection equation in a rectangular domain with
+uniform fluid velocity, constant diffusivity and a constant
+tracer source term. Neumann conditions are imposed on the
+channel walls, an inflow condition is imposed on the
+left-hand boundary, and the right-hand boundary remains open.
```
Comment (tkarna) [inline examples/point_discharge/point_discharge.py:9]: maybe mention here that this is a steady-state problem for which an analytical solution exists

**Review by tkarna — CHANGES_REQUESTED** [review]
Thanks for this! Looks good. As the script runs a convergence test this is more a test rather than an example? So would better fit under `test/tracerEq/`(?) rather than `examples/`. Otherwise only minor comments.

**Comment by jwallwork23** [comment]
Thanks for the review, @tkarna. I think I addressed all your comments.

**Review by tkarna — APPROVED** [review]
Thanks, looks good!

**Review by tkarna — APPROVED** [review]

**Pull request merged**